flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueWriters.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.iceberg.avro.ValueWriter;
+import org.apache.iceberg.avro.ValueWriters;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.TypeUtil;
+
+public class FlinkValueWriters {
+
+  private FlinkValueWriters() {}
+
+  static ValueWriter<StringData> strings() {
+    return StringWriter.INSTANCE;
+  }
+
+  static ValueWriter<byte[]> uuids() {
+    return ValueWriters.fixed(16);
+  }
+
+  static ValueWriter<Integer> timeMicros() {
+    return TimeMicrosWriter.INSTANCE;
+  }
+
+  static ValueWriter<TimestampData> timestampMicros() {
+    return TimestampMicrosWriter.INSTANCE;
+  }
+
+  static ValueWriter<DecimalData> decimal(int precision, int scale) {
+    return new DecimalWriter(precision, scale);
+  }
+
+  static <T> ValueWriter<ArrayData> array(ValueWriter<T> elementWriter, LogicalType elementType) {
+    return new ArrayWriter<>(elementWriter, elementType);
+  }
+
+  static <K, V> ValueWriter<MapData> arrayMap(
+      ValueWriter<K> keyWriter, LogicalType keyType, ValueWriter<V> valueWriter, LogicalType valueType) {
+    return new ArrayMapWriter<>(keyWriter, keyType, valueWriter, valueType);
+  }
+
+  static <K, V> ValueWriter<MapData> map(
+      ValueWriter<K> keyWriter, LogicalType keyType, ValueWriter<V> valueWriter, LogicalType valueType) {
+    return new MapWriter<>(keyWriter, keyType, valueWriter, valueType);
+  }
+
+  static ValueWriter<RowData> row(List<ValueWriter<?>> writers, List<LogicalType> types) {
+    return new RowWriter(writers, types);
+  }
+
+  private static class StringWriter implements ValueWriter<StringData> {
+    private static final StringWriter INSTANCE = new StringWriter();
+
+    private StringWriter() {
+    }
+
+    @Override
+    public void write(StringData s, Encoder encoder) throws IOException {
+      // toBytes is cheaper than Avro calling toString, which incurs encoding costs
+      encoder.writeString(new Utf8(s.toBytes()));
+    }
+  }
+
+  private static class DecimalWriter implements ValueWriter<DecimalData> {
+    private final int precision;
+    private final int scale;
+    private final int length;
+    private final ThreadLocal<byte[]> bytes;
+
+    private DecimalWriter(int precision, int scale) {
+      this.precision = precision;
+      this.scale = scale;
+      this.length = TypeUtil.decimalRequiredBytes(precision);
+      this.bytes = ThreadLocal.withInitial(() -> new byte[length]);
+    }
+
+    @Override
+    public void write(DecimalData d, Encoder encoder) throws IOException {
+      Preconditions.checkArgument(d.scale() == scale,
+          "Cannot write value as decimal(%s,%s), wrong scale: %s", precision, scale, d);
+      Preconditions.checkArgument(d.precision() <= precision,
+          "Cannot write value as decimal(%s,%s), too large: %s", precision, scale, d);
+
+      BigDecimal decimal = d.toBigDecimal();
+
+      byte fillByte = (byte) (decimal.signum() < 0 ? 0xFF : 0x00);
```

Comment:
Can we move this logic into a common `DecimalUtil` method? I think we have quite a few copies of it.

Reply:
Created https://github.com/apache/iceberg/pull/1265 for this.